rfd/0036/user-stories/jupiter-example-com.md

```diff
@@ -0,0 +1,31 @@
+
+Updates to the application and its content are done by building and staging new containers before deploying them to replace the existing containers.
+
+The application runs three instances of each tier in a single data center, but ideally it would have multiple data centers (perhaps with just one instance in each tier of each data center). The only application component that would need to peer across data centers is Consul (though this is uncommon and most apps would have other components that would need to peer across data centers).
+
+There are a small number of sensitive details that need to be carefully handled, including the SSL cert and key for Nginx, possibly TLS client and server certs for Consul, and a license key for the CMS. The application operators expect Mariposa to offer features to simplify that process.
+
+The code is hosted in GitHub and the images in a privately operated repository. The operators use Jenkins to automate the process of building new containers for each pull request and branch in the origin repo. After building the containers, Jenkins creates a new Mariposa project in which to stage it, or identifies and re-uses an existing project. The project names follow a predictable pattern based on the PR number of branch name (`<repo-name>-pr-<PR number>` or `<repo-name>-branch-<branch name>`), that allows the easy human and machine recognition of the relationship between projects in GitHub and Mariposa. The live, production project is named `<repo-name>-prod`, which cannot be matched by the naming logic the operators have defined in Jenkins.
+
+Despite the automation for continuous build and deploy of the staging projects, the deploys to the live site require manual promotion. After reviewing the relevant staging site and any logs generated by Jenkins, the operators merge a PR to master, triggering a new build of master that is tagged with the date and time of the build (the builds collect resources beyond the scope of the git repo, so the git SHA is not a consistent identifier). Jenkins stages that build in a project named `<repo-name>-branch-<branch name>` (using the same automation logic from above). After testing and previewing that final staging build, the operators can choose to manually deploy that build to the live site.
+
+At this point, we face a number of questions that need further exploration:
+
+1. How do we indicate which tag is the correct tag to run in production?
+2. How can we do that in a way that minimizes the risk of errors due to copypasta or other mistakes that are inherent in a manual process?
+3. Because tags are not immutable, we must assume that at some point the tag for an image in the registry will no longer be the same as the currently deployed image. How do we avoid accidental deploys if Triton/Mariposa scale up or replace an instance?
```

Comment:
A counterpoint on this idea is one of the processes we'd used at DF was to promote a build via tags:

- Jenkins builds `repo/service:latest` and tags it `repo/service:[githash]`
- to promote a build to a given environment tag `repo/service:[githash]` as `repo/service:[environment]`
- a given environment _always_ deploys the `repo/service:[environment]` tag

The exact details are probably not something we want to bake-in because folks will have different workflows. But we can have our tooling be allowed to update a service's deployment tag but not silently "set" that tag on top of an existing one.

Psuedo-code:
```
$ triton service set deploytag ${githash}
Ok.
$ triton service set deploytag ${newgithash}
Error: cannot overwrite a deploytag
$ triton service update deploytag ${githash} ${newgithash}
Ok.
```